packages/@aws-cdk/cdk/lib/synthesis.ts

```diff
@@ -0,0 +1,354 @@
+import fs = require('fs');
+import os = require('os');
+import path = require('path');
+import { collectRuntimeInformation } from './runtime-info';
+
+export interface ISynthesizable {
+  synthesize(session: ISynthesisSession): void;
+}
+
+export interface ISynthesisSession {
+  readonly store: ISessionStore;
+  readonly manifest: cxapi.AssemblyManifest;
+  addArtifact(id: string, droplet: cxapi.Artifact): void;
+  tryGetArtifact(id: string): cxapi.Artifact | undefined;
+}
+
+export interface SynthesisSessionOptions {
+  /**
+   * The file store used for this session.
+   */
+  store: ISessionStore;
+
+  /**
+   * Emit the legacy manifest (`cdk.out`) when the session is closed (alongside `manifest.json`).
+   * @default false
+   */
+  legacyManifest?: boolean;
+
+  /**
+   * Include runtime information (module versions) in manifest.
+   * @default true
+   */
+  runtimeInformation?: boolean;
+}
+
+export class SynthesisSession implements ISynthesisSession {
+  /**
+   * @returns true if `obj` implements `ISynthesizable`.
+   */
+  public static isSynthesizable(obj: any): obj is ISynthesizable {
+    return 'synthesize' in obj;
+  }
+
+  public readonly store: ISessionStore;
+
+  private readonly artifacts: { [id: string]: cxapi.Artifact } = { };
+  private _manifest?: cxapi.AssemblyManifest;
+  private readonly legacyManifest: boolean;
+  private readonly runtimeInfo: boolean;
+
+  constructor(options: SynthesisSessionOptions) {
+    this.store = options.store;
+    this.legacyManifest = options.legacyManifest !== undefined ? options.legacyManifest : false;
+    this.runtimeInfo = options.runtimeInformation !== undefined ? options.runtimeInformation : true;
+  }
+
+  public get manifest() {
+    if (!this._manifest) {
+      throw new Error(`Cannot read assembly manifest before the session has been finalized`);
+    }
+
+    return this._manifest;
+  }
+
+  public addArtifact(id: string, artifact: cxapi.Artifact): void {
+    cxapi.validateArtifact(artifact);
+    this.artifacts[id] = artifact;
+  }
+
+  public tryGetArtifact(id: string): cxapi.Artifact | undefined {
+    return this.artifacts[id];
+  }
+
+  public close(): cxapi.AssemblyManifest {
+    const manifest: cxapi.AssemblyManifest = this._manifest = {
+      version: cxapi.PROTO_RESPONSE_VERSION,
+      artifacts: this.artifacts,
+    };
+
+    if (this.runtimeInfo) {
+      manifest.runtime = collectRuntimeInformation();
+    }
+
+    this.store.writeFile(cxapi.MANIFEST_FILE, JSON.stringify(manifest, undefined, 2));
+
+    if (this.legacyManifest) {
+      const legacy: cxapi.SynthesizeResponse = {
+        ...manifest,
+        stacks: renderLegacyStacks(this.artifacts, this.store)
+      };
+
+      // render the legacy manifest (cdk.out) which also contains a "stacks" attribute with all the rendered stacks.
+      this.store.writeFile(cxapi.OUTFILE_NAME, JSON.stringify(legacy, undefined, 2));
+    }
+
+    return manifest;
+  }
+}
+
+export interface ISessionStore {
```

Comment:
How would I make a project directory available for a build? Would I symlink or manually copy each file into the store?

Reply:
I am thinking that project directories will be referenced directly in the Artifact's `build` field without copying them to the staging area.

I think that code generation and build will need to happen in-place in order to allow IDE integration to be sane. Otherwise, I can't see how users will be able to sanely consume any generated code.

You mentioned something yesterday about generating complete libraries and not code, and I wanted to loop back on that and tell you that this was not my intention when we talked about code generation last week, and I actually don't think it's very pragmatic to generate full libraries. All I was saying is that you should assume that users never edit or check in any generated code, so if you want users to consume something that you generated, they would do something like:

```ts
import { GeneratedClass } from '../cdk.generated';

// use GeneratedClass
```

Now, the CDK can generate it's code into `cdk.generated`.